File: C++/ising_wolff.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cstdlib>
#include <cmath>
#include <random>

const int N = 100000;  // Monte Carlo steps
const int L = 50;      // Square lattice size
const double T = 2.2;  // Temperature
const double p = 1.0 - exp(-2.0 / T);  // Bond probability

std::mt19937 rng(std::random_device{}());
std::uniform_real_distribution<> random_double(0.0, 1.0);

using Matrix = std::vector<std::vector<int>>;
using BoolMatrix = std::vector<std::vector<bool>>;

// Function declarations
void initial_state(Matrix &S);
void random_spin(int &i, int &j);
void cluster_formation(int i, int j, int Si, int &n_add, std::vector<std::pair<int, int>> &s_add, BoolMatrix &C, const Matrix &S);
void save_to_file(const Matrix &S, const std::string &filename);

int main() {
    Matrix S(L, std::vector<int>(L));  // Array of spins
    BoolMatrix C(L, std::vector<bool>(L, false));  // Clustered spins
    std::vector<std::pair<int, int>> s_add(4);  // Neighbors for cluster formation

    initial_state(S);
    save_to_file(S, "initial_state.txt");

    for (int it = 0; it < N; ++it) {
        std::fill(C.begin(), C.end(), std::vector<bool>(L, false));

        int i, j, Si;
        random_spin(i, j);
        C[i][j] = true;
        Si = S[i][j];

        int n_add = 0;
        cluster_formation(i, j, Si, n_add, s_add, C, S);

        while (n_add > 0) {
            std::vector<std::pair<int, int>> new_s_add;
            for (int ic = 0; ic < n_add; ++ic) {
                int x = s_add[ic].first;
                int y = s_add[ic].second;
                cluster_formation(x, y, Si, n_add, new_s_add, C, S);
            }
            s_add = new_s_add;
        }

        for (int x = 0; x < L; ++x) {
            for (int y = 0; y < L; ++y) {
                if (C[x][y]) {
                    S[x][y] = -S[x][y];
                }
            }
        }
    }

    save_to_file(S, "final_state.txt");
    return 0;
}

// Initialize spins randomly (-1 or 1)
void initial_state(Matrix &S) {
    for (int i = 0; i < L; ++i) {
        for (int j = 0; j < L; ++j) {
            S[i][j] = (random_double(rng) < 0.5) ? -1 : 1;
        }
    }
}

// Choose a random spin
void random_spin(int &i, int &j) {
    i = rng() % L;
    j = rng() % L;
}

// Cluster formation with periodic boundary conditions
void cluster_formation(int i, int j, int Si, int &n_add, std::vector<std::pair<int, int>> &s_add, BoolMatrix &C, const Matrix &S) {
    int ip = (i + 1) % L, im = (i - 1 + L) % L;
    int jp = (j + 1) % L, jm = (j - 1 + L) % L;

    n_add = 0;

    if (S[ip][j] == Si && random_double(rng) < p && !C[ip][j]) {
        s_add.emplace_back(ip, j);
        C[ip][j] = true;
        ++n_add;
    }
    if (S[im][j] == Si && random_double(rng) < p && !C[im][j]) {
        s_add.emplace_back(im, j);
        C[im][j] = true;
        ++n_add;
    }
    if (S[i][jp] == Si && random_double(rng) < p && !C[i][jp]) {
        s_add.emplace_back(i, jp);
        C[i][jp] = true;
        ++n_add;
    }
    if (S[i][jm] == Si && random_double(rng) < p && !C[i][jm]) {
        s_add.emplace_back(i, jm);
        C[i][jm] = true;
        ++n_add;
    }
}

// Save lattice state to file
void save_to_file(const Matrix &S, const std::string &filename) {
    std::ofstream file(filename);
    for (const auto &row : S) {
        for (const auto &val : row) {
            file << val << " ";
        }
        file << "\n";
    }
}

File: fortran/ising_wolff.f90
program ising_wolff
    implicit none
    
    !------------------------------------------------------------------------!
    !                               parameters                               !
    !------------------------------------------------------------------------!

    integer, parameter :: N = 1e3                 ! Monte Carlo steps
    integer, parameter :: L = 50,nx = L, ny = L   ! square lattice size
    real(8), parameter :: T = 2.0d0              ! temperature
    real(8), parameter :: p = 1.0d0-exp(-2.0d0/T) ! bonds probability
    integer, parameter :: f = 1e0                 ! saving data frecuency
    
    !------------------------------------------------------------------------!
    !                               variables                                !
    !------------------------------------------------------------------------!

    integer :: it, i, j, Si, n_add, ic, E, M, ip, jp
    real(8) :: r
    integer, dimension(nx,nx) :: S ! array of spins
    logical, dimension(nx,ny) :: C ! array of clustered spins
    integer, dimension(2,nx*ny) :: s_add

    !------------------------------------------------------------------------!
    !                        Monte Carlo Wolff Method                        !
    !------------------------------------------------------------------------!

    call initial_state(S) ! asign random values to the spins (1 or -1)

    open (unit = 100, file = "state_evolution.txt", status = "replace")
    open (unit = 101, file = "magnetization.txt", status = "replace")
    open (unit = 102, file = "energy.txt", status = "replace")
    E = 0
    M = 0
    do it = 1,N
        C = .FALSE.
        call random_spin(i,j) ! randomly choose a spin
        C(i,j) = .TRUE.
        Si = S(i,j) ! value of the choosen spin
        n_add = 1
        s_add(1,1) = i
        s_add(2,1) = j
        
        do ic = 1,n_add
            call cluster_formation(Si,s_add(1,ic),s_add(2,ic),n_add,s_add,C)
        end do

        ! cluster flip
        do i = 1,nx
            do j = 1,ny
                if (C(i,j)) S(i,j) = -S(i,j)
            end do 
        end do 

        ! save the data
        if (mod(it,f) == 0) then
            M = 0
            E = 0
            do i = 1,nx
                ip = mod(i,nx)+1
                do j = 1,ny
                    jp = mod(j,ny)+1
                    M = M+S(i,j)
                    E = E-(S(i,j)*(S(ip,j)+S(i,jp)))
                    write(100,*) S(i,j)
                end do 
            end do
            write(101,*) M
            write(102,*) E/2
        end if
    end do
    close(100)
    close(101)
    close(102)

    !------------------------------------------------------------------------!
    !                        functions and subroutines                       !
    !------------------------------------------------------------------------!

    contains

    ! random initial state
    subroutine initial_state(S)
        integer, intent(out) :: S(nx,ny)

        open (unit = 100, file = "parameters.txt", status = "replace")
        open (unit = 101, file = "initial_state.txt", status = "replace")
            write(100,*) L, N, f
            do i = 1,L
                do j = 1,L
                    call random_number(r)
                    S(i,j) = 2*int(r+0.5d0)-1
                    write(101,*) S(i,j)
                end do
            end do
        close(100)
        close(101)

    end subroutine initial_state

    ! random initial state
    subroutine random_spin(i,j)
        integer, intent(out) :: i, j

        call random_number(r)
        i = 1+int(r*nx)
        call random_number(r)
        j = 1+int(r*ny)

    end subroutine random_spin

    ! neighbors with periodic boudary conditions for cluster formation
    subroutine cluster_formation(Si,i,j,n_add,s_add,C)
        integer, intent(in) :: Si, i, j
        integer, intent(inout) :: s_add(2,nx*ny), n_add
        logical, intent(inout) :: C(nx,ny)
        integer :: ip, im, jp, jm

        ip = mod(i,nx)+1
        im = mod(i-2,nx)+1
        jp = mod(j,ny)+1
        jm = mod(j-2,ny)+1

        call random_number(r)
        if (S(ip,j) == Si .AND. r < p .AND. .NOT. C(ip,j)) then
            n_add = n_add+1
            s_add(1,n_add) = ip
            s_add(2,n_add) = j
            C(ip,j) = .TRUE.
        end if
        call random_number(r)
        if (S(im,j) == Si .AND. r < p .AND. .NOT. C(im,j)) then
            n_add = n_add+1
            s_add(1,n_add) = im
            s_add(2,n_add) = j
            C(im,j) = .TRUE.
        end if
        call random_number(r)
        if (S(i,jp) == Si .AND. r < p .AND. .NOT. C(i,jp)) then
            n_add = n_add+1
            s_add(1,n_add) = i
            s_add(2,n_add) = jp
            C(i,jp) = .TRUE.
        end if
        call random_number(r)
        if (S(i,jm) == Si .AND. r < p .AND. .NOT. C(i,jm)) then
            n_add = n_add+1
            s_add(1,n_add) = i
            s_add(2,n_add) = jm
            C(i,jm) = .TRUE.
        end if

    end subroutine cluster_formation

end program ising_wolff